<div style="border:solid purple 5px; padding: 20px"> 

<b>Кирилл, привет 🖖!</b>
    
Меня зовут Даша Абраменкова и я твой тимлид.
Для упрощения коммуникации мои пометки будут в отдельных Markdown ячейках. <b>Пожалуйста, не удаляй и не меняй их</b> — так будет проще отслеживать внесенные изменения. К слову о доработках. Будет намного удобнее, если о сделанных изменениях ты так же сообщишь мне в отдельной Markdown ячейке (это позволит принять проект куда быстрее). 
    
<hr>    
Несколько  слов об оформлении. Мои комментарии будут оформлены тремя цветами:

<div class="alert alert-success">
<b> 🤌  Комментарий тимлида</b>   — зеленый комментарий означает, что эта часть работы выполнена отлично;</div>
    
<div class="alert alert-warning">
<b>🕵️‍♀️  Комментарий тимлида </b> — желтый цвет укажет на хорошо проделанную работу и предложит (хоть это и необязательно) немного доработать текущий этап проекта;</div>
    
<div class="alert alert-danger">
<b> 👨‍🚒  Комментарий тимлида</b> — красный комментарий означает, что некоторые аспекты придется изменить, так как без их исправления работу принять я не смогу.</div> 
    
    
<div class="alert alert-info">
<b>🧑‍🎓 Комментарий джуна </b> — если хочешь, то можешь использовать синий цвет для своих комментариев, скопировав разметку из этой ячейки</div>    



</div>      

In [1]:
from IPython.display import Image

# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


<img src="https://concrete-web-bad.notion.site/image/https%3A%2F%2Fs3-us-west-2.amazonaws.com%2Fsecure.notion-static.com%2F069818d1-0e5c-4d87-a461-0de584ab9c33%2FUntitled_(33).png?id=bd53c8db-b4fd-49eb-8cc9-572ebb3c9163&table=block&spaceId=9e4bd47b-c6e6-4ca3-bcee-279794b47315&width=1640&userId=&cache=v2" alt="drawing" width="700"/>


### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Как выполнить задание?

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

### Доступ к базе данных

Следуйте инструкции, приложенной ниже

[instructions_for_connecting_to_the_database.pdf](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/bac02c0e-9489-4458-acff-4cd0ad912b46/instructions_for_connecting_to_the_database.pdf)

### Примечания

- Не забывайте про функции, они значительно облегчат жизнь при выполнении запросов;
- Получите ответы, используя только язык запросов SQL. Pandas допускается только для хранения и вывода результатов.

# Содержание проекта
* 1. [Опишите цели исследования](#1)
* 2. [Исследуйте таблицы — выведите первые строки](#2)
* 3. [задания](#3)
    - Задание 1: Посчитайте, сколько книг вышло после 1 января 2000 года;
    - Задание 2: Для каждой книги посчитайте количество обзоров и среднюю оценку;
    - Задание 3: Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
    - Задание 4: Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
    - Задание 5: Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
* 4. [Выведите результат каждого запроса в тетрадке](#4)
* 5. [Опишите выводы по каждой из решённых задач](#5)

# 1. Цели исследования
- Наша компания купила крупный сервис для чтения книг по подписке. 
- Первая задача как аналитика — проанализировать базу данных.
    - В ней информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# 2. Исследование таблицы <a id=1></a>

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

<div class="alert alert-warning">
<b>🕵️‍♀️  Комментарий тимлида </b>


Когда и импорт библиотек, и импорт файла, и создание подключения происходят в одной ячейке, то изменения в любом из этих шагов, безусловно, ведут к перезаписи всех остальных.
    
    
Например, ты захочешь расширить число используемых библиотек, но для этого заново придется перезаписывать переменную, а значит пересчитывать ноутбук. 
    
    
Чтобы этого избежать, лучше сразу разделять логически разные вещи:
    
— импорт библиотек   
— загрузка файлов    
— рабочий код   
— и т.д.
    </div>

#### Выведем первые строки таблиц

In [3]:
# Упростим, напишем функцию для обработки запросов SQL
def sql(query):
    return pd.io.sql.read_sql(query, con = engine) 

def info_all(data):
    '''Функция вываедет все рядовые показатели при предобработке данных'''
    display(data.head(5))
    print()
    print(data.info())
    print()
    print(f'Кол-во дублей {data.duplicated().sum()} строк')
    print()
    print(f'Кол-во пропусков')
    print(data.isna().sum())
    print()
    display(data.describe())
    print('-----------------------------------------------------------------------------------------')

<div class="alert alert-warning">
<b>🕵️‍♀️  Комментарий тимлида </b>


Круто, что ты автоматизируешь код — это позволяет потратить врем всего один раз, вместо того, чтобы копипастить код весь проект. Но я бы еще рекомендовала ограничивать запрос командой `limit`, если речь идет о знакомстве с даными. Ведь если данных действительно много (а не как у нас), то попытка выгрузить гигабайты информации может попросту поломать сервер, забив память :)
    
    
<a href = "https://dzone.com/articles/why-you-should-not-use-select-in-sql-query-1"> 7 причин, почему запрос в таком виде — плохая идея </a>
    </div>

In [4]:
sheets_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for i in sheets_list:
    print()
    print('Таблица: {}'.format(i))
    y = sql('''SELECT * FROM {}'''.format(i))
    info_all(y)


Таблица: books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None

Кол-во дублей 0 строк

Кол-во пропусков
book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64



book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

-----------------------------------------------------------------------------------------

Таблица: authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None

Кол-во дублей 0 строк

Кол-во пропусков
author_id    0
author       0
dtype: int64



author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

-----------------------------------------------------------------------------------------

Таблица: publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None

Кол-во дублей 0 строк

Кол-во пропусков
publisher_id    0
publisher       0
dtype: int64



publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

-----------------------------------------------------------------------------------------

Таблица: ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None

Кол-во дублей 0 строк

Кол-во пропусков
rating_id    0
book_id      0
username     0
rating       0
dtype: int64



rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

-----------------------------------------------------------------------------------------

Таблица: reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None

Кол-во дублей 0 строк

Кол-во пропусков
review_id    0
book_id      0
username     0
text         0
dtype: int64



review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

-----------------------------------------------------------------------------------------


### C Данными все впорядке.
- Отсутствуют дубликаты
- Отсутствуют пропуски
- Типы данных соответствуют, кроме `publication_date` в таблице `books`

<a id=3></a>
# 3. Выполнение заданий

###  Задание 1: Посчитайте, сколько книг вышло после 1 января 2000 года;


In [5]:
print('После 1 января 2000 года, вышло {} книг'.format(sql(
    """SELECT COUNT(book_id) FROM books WHERE publication_date > '2000-01-01'""").loc[0][0]
                                                      )
     )

После 1 января 2000 года, вышло 819 книг


<div class="alert alert-success">
<b> 🤌  Комментарий тимлида </b>  
    
Отлично сработано
    
</div>

### Задание 2: Для каждой книги посчитаем количество обзоров и среднюю оценку;

In [6]:
reviews_rating = sql("""

SELECT 
    books.title,
    AVG(ratings.rating),
    COUNT(DISTINCT reviews.text)
FROM books 
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY
    books.title

""")
reviews_rating

title       avg  count
0                                         'Salem's Lot  3.666667      2
1                   1 000 Places to See Before You Die  2.500000      1
2    13 Little Blue Envelopes (Little Blue Envelope...  4.666667      3
3    1491: New Revelations of the Americas Before C...  4.500000      2
4                                                 1776  4.000000      4
..                                                 ...       ...    ...
994           Wyrd Sisters (Discworld  #6; Witches #2)  3.666667      3
995                        Xenocide (Ender's Saga  #3)  3.400000      3
996                                    Year of Wonders  3.200000      4
997                        You Suck (A Love Story  #2)  4.500000      2
998  Zen and the Art of Motorcycle Maintenance: An ...  3.833333      4

[999 rows x 3 columns]

<div class="alert alert-success">
<b> 🤌  Комментарий тимлида </b>  
    
correct ✅
    
</div>

### Задание 3: Определить издательство, которое выпустило наибольшее число книг толще 50 страниц
- так мы исключим из анализа брошюры;

In [7]:
sql("""

SELECT 
    p.publisher,
    COUNT(b.title) AS count
FROM 
    books AS b
LEFT JOIN 
    publishers AS p ON p.publisher_id = b.publisher_id
WHERE 
    b.num_pages > 50
GROUP BY 
    p.publisher
ORDER BY 
    count DESC
LIMIT (1)

""")

publisher  count
0  Penguin Books     42

- Наибольшее кол-во книг (толще 50 страниц) выпустило издательство `Penguin Books`, в колличестве 42 штук. 

<div class="alert alert-success">
<b> 🤌  Комментарий тимлида </b>  
    
Совершенно верно
    
    
![](https://i.gifer.com/4kj1.gif)
    
    
</div>

### Задание 4: Определите автора с самой высокой средней оценкой книг
- Учитываем только книги с 50 и более оценками;

In [8]:
sql('''
SELECT
    y.author,
    AVG(y.avg_rat) as avg_rat
FROM
    (SELECT 
        b.title,
        a.author,
        AVG(r.rating) AS avg_rat
    FROM 
        books AS b
    LEFT JOIN 
        authors AS a ON a.author_id = b.author_id
    LEFT JOIN 
        ratings AS r ON r.book_id = b.book_id
    GROUP BY 
        b.title,
        a.author
    HAVING    
        COUNT(r.rating) >= 50) AS y
GROUP BY 
    y.author
ORDER BY
    avg_rat DESC
LIMIT (1)  

''')

author   avg_rat
0  J.K. Rowling/Mary GrandPré  4.283844

- Самая высока средняя оценка у автора `J.K. Rowling/Mary GrandPré` с рейтингом `4.28` 

<div class="alert alert-success">
<b> 🤌  Комментарий тимлида </b>  
    
Это сложна задача, а ты ее выполнил с первого раза, супер
</div>

### Задание 5: Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [9]:
sql('''
SELECT ROUND(AVG(y.count),2)
FROM
    (SELECT 
        r.username,
        COUNT(r.text)

    FROM
        (SELECT 
            username,
            COUNT(rating_id) as count_rat
        FROM 
            ratings
        GROUP BY
            username
        HAVING COUNT(rating_id) > 50) AS count_rat
    INNER JOIN
        reviews AS r ON r.username = count_rat.username

    GROUP BY 
        r.username) as y
''')

round
0  24.33

- 24 обзора в среднем написали пользователи, которые оставили более 50 оценок

<div class="alert alert-success">
<b> 🤌  Комментарий тимлида </b>  
    
Все так
    
</div>

# Общий вывод:
- После 1 января 2000 года, вышло 819 книг
- Дата фрейм `reviews_rating` содержит по каждой книге кол-во обхоров и среднюю оценку книги
- Наибольшее кол-во книг (толще 50 страниц) выпустило издательство Penguin Books, в колличестве 42 штук.
- Самая высока средняя оценка у автора J.K. Rowling/Mary GrandPré с рейтингом 4.28
- 24 обзора в среднем написали пользователи, которые оставили более 50 оценок

<div class="alert alert-success">
<b> 🤌  Комментарий тимлида </b>  
    
Как замечательно: `SQL` дался тебе с первого раза. Что ж, значит позволь поздравить с завершением финального спринта 
    
![](https://i.gifer.com/Be.gif)
    
</div>